validate/validate.go

```diff
@@ -0,0 +1,610 @@
+package validate
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"reflect"
+	"strings"
+	"unicode"
+	"unicode/utf8"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/blang/semver"
+	rspec "github.com/opencontainers/runtime-spec/specs-go"
+)
+
+var (
+	defaultRlimits = []string{
+		"RLIMIT_CPU",
+		"RLIMIT_FSIZE",
+		"RLIMIT_DATA",
+		"RLIMIT_STACK",
+		"RLIMIT_CORE",
+		"RLIMIT_RSS",
+		"RLIMIT_NPROC",
+		"RLIMIT_NOFILE",
+		"RLIMIT_MEMLOCK",
+		"RLIMIT_AS",
+		"RLIMIT_LOCKS",
+		"RLIMIT_SIGPENDING",
+		"RLIMIT_MSGQUEUE",
+		"RLIMIT_NICE",
+		"RLIMIT_RTPRIO",
+		"RLIMIT_RTTIME",
+	}
+	defaultCaps = []string{
+		"CAP_CHOWN",
+		"CAP_DAC_OVERRIDE",
+		"CAP_FSETID",
+		"CAP_FOWNER",
+		"CAP_MKNOD",
+		"CAP_NET_RAW",
+		"CAP_SETGID",
+		"CAP_SETUID",
+		"CAP_SETFCAP",
+		"CAP_SETPCAP",
+		"CAP_NET_BIND_SERVICE",
+		"CAP_SYS_CHROOT",
+		"CAP_KILL",
+		"CAP_AUDIT_WRITE",
+	}
+)
+
+type Validator struct {
+	Spec         *rspec.Spec
+	RootfsPath   string
+	HostSpecific bool
+}
+
+func NewValidator(spec *rspec.Spec, rootfsPath string, hostSpecific bool) Validator {
+	return Validator{Spec: spec, RootfsPath: rootfsPath, HostSpecific: hostSpecific}
+}
+
+func NewValidatorFromPath(bundlePath string, hostSpecific bool) (Validator, error) {
+	if bundlePath == "" {
+		return Validator{}, fmt.Errorf("Bundle path shouldn't be empty")
+	}
+
+	if _, err := os.Stat(bundlePath); err != nil {
+		return Validator{}, err
+	}
+
+	configPath := path.Join(bundlePath, "config.json")
+	content, err := ioutil.ReadFile(configPath)
+	if err != nil {
+		return Validator{}, err
+	}
+	if !utf8.Valid(content) {
+		return Validator{}, fmt.Errorf("%q is not encoded in UTF-8", configPath)
+	}
+	var spec rspec.Spec
+	if err = json.Unmarshal(content, &spec); err != nil {
+		return Validator{}, err
+	}
+
+	rootfsPath := path.Join(bundlePath, spec.Root.Path)
+	if fi, err := os.Stat(rootfsPath); err != nil {
+		return Validator{}, fmt.Errorf("Cannot find the root path %q", rootfsPath)
+	} else if !fi.IsDir() {
+		return Validator{}, fmt.Errorf("The root path %q is not a directory.", rootfsPath)
+	}
```

Comment:
This “does the rootfs exist” check should be a separate check in `Validator.CheckAll()`